File: tests/variable/subst/variable_subst.ml
open Optitrust
open Prelude

let _ = Flags.check_validity := true
let _ = Flags.recompute_resources_between_steps := true

let _ = Run.script_cpp (fun _ ->
  let x = trm_find_var "x" [] in
  let (y, _) = find_var "y" [] in
  (* TODO: (expr "2 + x") *)
  !! Variable_basic.subst ~subst:y ~put:(trm_add x (trm_int 2)) [cVarDef "z"];
  !! Variable_basic.subst ~subst:y ~put:(trm_int 5) [cOr [[cVarDef "a"]; [cWriteVar "z"]]];
)

File: tests/variable/subst/variable_subst.cpp
#include <optitrust.h>

void f() {
  __pure();

  const int x = 3; /* TODO: non const x */
  const int y = x + 2;
  int z = y + y;
  int a = y + 3;
  z = y + 2;
  z = y + y;
}
